### PR TITLE
bitwarden: 2023.3.2 -> 2023.4.0; backport update to build against Node 18

### PR DESCRIPTION
**Pull request by amarshall**
###### Description of changes

- Backport patch from Bitwarden master to achieve this as they have done
  (unreleased) upgrade 16→18, and have several other changes along with
  it. We want this now because Node 16 is being marked insecure soon for
  NixOS 23.05; see https://github.com/NixOS/nixpkgs/pull/229910.
- These changes should be in the next release in a few weeks
- `npm bin` no longer exists, use `npm exec` instead

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/security/bitwarden/default.nix [hunk 1]
```diff
@@ -22,23 +23,32 @@ let
...
   src = fetchFromGitHub {
     owner = "bitwarden";
     repo = "clients";
     rev = "desktop-v${version}";
-    sha256 = "sha256-KQDM7XDUA+yRv8y1K//rMCs4J36df42RVsiAXazJeYQ=";
+    sha256 = "sha256-TTKDl6Py3k+fAy/kcyiMbAAKQdhVnZTyRXV8D/VpKBE=";
   };
 
+  patches = [
+    (fetchurl {
```
Comment (SuperSandro2000) [inline pkgs/tools/security/bitwarden/default.nix:38]:
```suggestion
    (fetchpatch {
```
and please update the patch

File: pkgs/tools/security/bitwarden/default.nix [hunk 2]
```diff
@@ -22,23 +23,32 @@ let
...
-    inherit src version;
-    sourceRoot = "source/apps/desktop/desktop_native";
-    cargoSha256 = "sha256-XsAmVYWPPnY0cgBzpO2aWx/fh85fKr8kMO98cDMzOKk=";
+    inherit src version patches;
+    cargoSha256 = "sha256-VW9DmSh9jvqFCZjH1SAYkydSGjXSVEbv4CmtoJBiw5Y=";
 
-    patchFlags = [ "-p4" ];
+    postPatch = ''
+      cd apps/desktop/desktop_native
+    '';
```
Comment (SuperSandro2000) [inline pkgs/tools/security/bitwarden/default.nix:51]: Please use sourceRoot

File: pkgs/tools/security/bitwarden/default.nix [hunk 3]
```diff
@@ -102,6 +112,15 @@ buildNpmPackage' {
     python3
   ];
 
+  postPatch = ''
+    expectedElectronVersionMajor=$(jq --raw-output '.devDependencies.electron' < package.json | grep -E --only-matching '^[0-9]+')
+    if [[ $expectedElectronVersionMajor != ${lib.escapeShellArg (lib.versions.major electron.version)} ]]; then
+      echo 'ERROR: electron version mismatch'
+      exit 1
+    fi
+    unset expectedElectronVersionMajor
```
Comment (SuperSandro2000) [inline pkgs/tools/security/bitwarden/default.nix:121]:
```suggestion
    if [[ $(jq --raw-output '.devDependencies.electron' < package.json | grep -E --only-matching '^[0-9]+') != ${lib.escapeShellArg (lib.versions.major electron.version)} ]]; then
      echo 'ERROR: electron version mismatch'
      exit 1
    fi
```

File: pkgs/tools/security/bitwarden/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/tools/security/bitwarden/default.nix:51]: I skipped over that. Then as you figured out, we don't want to set sourceRoot.

File: pkgs/tools/security/bitwarden/default.nix [hunk 4]
```diff
@@ -22,23 +23,32 @@ let
...
   src = fetchFromGitHub {
     owner = "bitwarden";
     repo = "clients";
     rev = "desktop-v${version}";
-    sha256 = "sha256-KQDM7XDUA+yRv8y1K//rMCs4J36df42RVsiAXazJeYQ=";
+    sha256 = "sha256-TTKDl6Py3k+fAy/kcyiMbAAKQdhVnZTyRXV8D/VpKBE=";
   };
 
+  patches = [
+    (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/tools/security/bitwarden/default.nix:38]:
```suggestion
    # Bump electron to 24 and node to 18
    (fetchpatch {
```

File: pkgs/tools/security/bitwarden/default.nix [hunk 5]
```diff
@@ -22,23 +23,32 @@ let
...
   desktop-native = rustPlatform.buildRustPackage {
     pname = "bitwarden-desktop-native";
-    inherit src version;
-    sourceRoot = "source/apps/desktop/desktop_native";
-    cargoSha256 = "sha256-XsAmVYWPPnY0cgBzpO2aWx/fh85fKr8kMO98cDMzOKk=";
+    inherit src version patches;
+    cargoSha256 = "sha256-VW9DmSh9jvqFCZjH1SAYkydSGjXSVEbv4CmtoJBiw5Y=";
 
-    patchFlags = [ "-p4" ];
+    postPatch = ''
```
Comment (SuperSandro2000) [inline pkgs/tools/security/bitwarden/default.nix:49]: we could move that to preConfigure or preBuild but it really does not matter

File: pkgs/tools/security/bitwarden/default.nix [hunk 6]
```diff
@@ -102,6 +112,13 @@ buildNpmPackage' {
...
   ];
 
+  postPatch = ''
+    if [[ $(jq --raw-output '.devDependencies.electron' < package.json | grep -E --only-matching '^[0-9]+') != ${lib.escapeShellArg (lib.versions.major electron.version)} ]]; then
+      echo 'ERROR: electron version mismatch'
+      exit 1
+    fi
+  '';
+
   preBuild = ''
```
Comment (SuperSandro2000) [inline pkgs/tools/security/bitwarden/default.nix:122]: I think we want to move that to a later step in the build to not break the vendoring

```
patching sources
/nix/store/f77w5n01j0s8qlhniw79ms0pv1dpyh80-stdenv-linux/setup: line 119: jq: command not found
ERROR: electron version mismatch
configuring
error: builder for '/nix/store/i9cxlz7isqj78bzbva74a9a4fikf36q4-bitwarden-2023.4.0-npm-deps.drv' failed with exit code 1;
       last 10 log lines:
       > Hunk #7 succeeded at 39152 (offset 1 line).
       > Hunk #8 succeeded at 71235 (offset 2 lines).
       > Hunk #9 succeeded at 71251 (offset 2 lines).
       > Hunk #10 succeeded at 71274 (offset 2 lines).
       > Hunk #11 succeeded at 71664 (offset 2 lines).
       > Hunk #12 succeeded at 74106 (offset 2 lines).
       > patching file package.json
       > patching file package.json
       > /nix/store/f77w5n01j0s8qlhniw79ms0pv1dpyh80-stdenv-linux/setup: line 119: jq: command not found
       > ERROR: electron version mismatch
       For full logs, run 'nix log /nix/store/i9cxlz7isqj78bzbva74a9a4fikf36q4-bitwarden-2023.4.0-npm-deps.drv'.
no configure script, doing nothing
building
ERROR: The Cargo.lock file doesn't exist
Cargo.lock is needed to make sure that cargoHash/cargoSha256 doesn't change
when the registry is updated.
```

**Comment by mweinelt** [comment]
Tested and works for me.

**Review by SuperSandro2000 — APPROVED** [review]
tested locally and works

**Pull request merged**